### PR TITLE
Feature/pi 2158 mas get unpaid work time

### PR DESCRIPTION
**Pull request by achimber-moj**

### CODE REVIEW

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt [hunk 1]
```diff
@@ -116,6 +119,22 @@ class SentenceService(
         return null
     }
 
+    fun getUnpaidWorkTime(requirementType: String, eventId: Long, eventNumber: String): String? {
+        if (requirementType.equals("W", true)) {
+            val durationInMinutes: Long = upwAppointmentRepository.calculateUnpaidTimeWorked(eventId, eventNumber)
+            durationInMinutes.let {
+                val durationInHours = it / 60
```
Comment (pmcphee77) [inline projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt:126]: do we need to think about rounding here?

**Review by pmcphee77 — CHANGES_REQUESTED** [review]

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt (hunk 1 above)
Comment (achimber-moj) [reply projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt:126]: will change the logic for this one

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/api/model/sentence/Requirement.kt [hunk 2]
```diff
@@ -13,6 +13,7 @@ data class Requirement(
     val description: String,
     val length: Long?,
     val lengthUnitValue: String?,
+    val progress: String?,
```
Comment (marcus-bcl) [inline projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/api/model/sentence/Requirement.kt:16]: Could we call this something more UPW-specific?  E.g. `unpaidWorkProgress`?

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/sentence/entity/UpwAppointment.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                SELECT ua.MINUTES_CREDITED AS duration
+                FROM UPW_APPOINTMENT ua
+                JOIN OFFENDER o 
+                ON o.OFFENDER_ID = ua.OFFENDER_ID 
+                JOIN EVENT e 
+                ON e.OFFENDER_ID = o.OFFENDER_ID 
+                WHERE e.EVENT_ID = :id
+                AND e.EVENT_NUMBER = :eventNumber
+                AND ua.ATTENDED = 'Y'
+                AND ua.SOFT_DELETED = 0)
```
Comment (marcus-bcl) [inline projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/sentence/entity/UpwAppointment.kt:41]: I'm not sure you need this sub-select.  Also, this looks like it'll get any UPW appointments for the offender, even if the appointments are for other events?  You might be able to link to the sentence (disposal) via the upw_details, something like:

```suggestion
            select sum(minutes_credited) 
            from upw_appointment 
            join upw_details on upw_details.upw_details_id = upw_appointment.upw_details_id
            join disposal on disposal.disposal_id = upw_details.disposal_id
            where disposal.event_id = :eventId
            and upw_appointment.attended = 'Y'
            and upw_appointment.soft_deleted = 0
```

However at that point, are we sure we should be returning this as part of the Requirement model - if it's actually linked to the Sentence?

Comment (achimber-moj) [reply projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/sentence/entity/UpwAppointment.kt:41]: sql updated.  i am happy to leave the model at the moment

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt (hunk 1 above)
Comment (achimber-moj) [reply projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt]: updated logic

**Pull request closed without merging**